### PR TITLE
ENH: Tiled QR wrappers for scipy.linalg.lapack

### LINKED ISSUE #10323: Adding more low-level LAPACK wrappers
I would like to be able to make use of the following LAPACK routines from `scipy.linalg.lapack` (which currently have Cython wrappers provided in `scipy.linalg.cython_lapack`): `geqrt`, `gemqrt`, `tpqrt` and `tpmqrt` for tiled QR factorisation. I suggest the following signatures (with work and info as necessary):

```
?geqrt(nb, a, overwrite_a) -> (a, t)
?gemqrt(side, trans, v, t, c, overwrite_c) -> (c)
?tpqrt(l, nb, a, b, overwrite_a, overwrite_b) -> (a, b, t)
?tpmqrt(side, trans, v, t, a, b, overwrite_a, overwrite_b) -> (a, b)
```

Since QR on F-contiguous is equivalent to LQ on C-contiguous, in principle the LQ factorisation equivalents `gelqt`, `gemlqt`, `tplqt`, `tpmlqt` could be called on C-contiguous arrays to avoid a transposition at the high level. However, these routines do not have the Cython wrappers, which I assume is due to their relatively recent introduction to LAPACK.

I am happy to work on a PR for getting at least the four tiled QR routines going, but I will need some guidance with what's involved to do so.

### PR DESCRIPTION
**Pull request by AidanGG**
As per https://github.com/scipy/scipy/issues/10323 I have begun working on the wrappers. I believe the wrappers and documentation are complete.

I still need to write some tests (but they do appear to function according to my brief, informal testing).

Closes #10323

### CODE REVIEW

**Comment by AidanGG** [comment]
@ilayn I believe this is ready to be looked at, but there are some CI issues with the documentation? How would I fix this?

**Comment by ilayn** [comment]
Great. Thanks for the time and effort. I will check further today more carefully. 

One of the travis failures is real though for float32 dtype. The fft stuff you can ignore for now; it is a new feature and CIs didn't refreshed apparently

**Comment by AidanGG** [comment]
Thanks, my test seems to fail on Travis with Python 3.5. I will investigate this locally with a new virtualenv.

File: scipy/linalg/flapack_other.pyf.src [hunk 1]
```diff
@@ -401,6 +401,116 @@ subroutine <prefix2c>unmqr(side,trans,m,n,k,a,lda,tau,c,ldc,work,lwork,info)
...
+    !
+    ! Multiplies a general matrix C by the orthogonal matrix Q defined by the
+    ! elementary reflectors stored in matrix V and the upper triangular block
+    ! reflectors in matrix T. C may be multiplied by Q, its transpose (for real
+    ! matrices), or its adjoint (for complex matrices) from the left or right.
+
+    callstatement (*f2py_func)(side,trans,&m,&n,&k,&nb,v,&ldv,t,&ldt,c,&ldc,work,&info)
+    callprotoargument char*,char*,int*,int*,int*,int*,<ctype>*,int*,<ctype>*,int*,<ctype>*,int*,<ctype>*,int*
+
+    character intent(in),check(*side=='L'||*side=='R'):: side
```
Comment (ilayn) [inline scipy/linalg/flapack_other.pyf.src]: These chars best be optional with default values. I know we are not perfectly consistent but we are trying to fix them as we wrap new functions into these files.

Comment (ilayn) [reply scipy/linalg/flapack_other.pyf.src]: Having said that we also need a simple test that gives wrong chars as arguments and we expect them to fail.

File: scipy/linalg/flapack_other.pyf.src [hunk 2]
```diff
@@ -401,6 +401,116 @@ subroutine <prefix2c>unmqr(side,trans,m,n,k,a,lda,tau,c,ldc,work,lwork,info)
...
+    ! elementary reflectors stored in the pentagonal matrix V and the upper
+    ! triangular block reflectors in matrix T. L is the order of the trapezoidal
+    ! part of matrix V. Matrix C consists of blocks A and B, and may be
+    ! multiplied by Q, its transpose (for real matrices), or its adjoint (for
+    ! complex matrices) from the left or right.
+
+    callstatement (*f2py_func)(side,trans,&m,&n,&k,&l,&nb,v,&ldv,t,&ldt,a,&lda,b,&ldb,work,&info)
+    callprotoargument char*,char*,int*,int*,int*,int*,int*,<ctype>*,int*,<ctype>*,int*,<ctype>*,int*,<ctype>*,int*,<ctype>*,int*
+
+    character intent(in),check(*side=='L'||*side=='R'):: side
```
Comment (ilayn) [inline scipy/linalg/flapack_other.pyf.src]: same as above

**Review by ilayn — CHANGES_REQUESTED** [review]
Other than the minor issue, I don't have any other issues.

File: scipy/linalg/flapack_other.pyf.src (hunk 1 above)
Comment (AidanGG) [reply scipy/linalg/flapack_other.pyf.src]: Great, I'm assuming `trans` should also have the default value `'N'`?

Comment (ilayn) [reply scipy/linalg/flapack_other.pyf.src]: Yes indeed. And the side is L

Comment (AidanGG) [reply scipy/linalg/flapack_other.pyf.src]: Alright, done, with tests, in the latest commit.

**Comment by ilayn** [comment]
The CircleCI stuff is irrelevant but in Travis run I think you are hitting the same issue that I had in #8965 

The documentation is getting choked on the first unmatched `*` character of the `size` while trying to write the parameter bounds. You can use `side[0]` instead to escape from it.

**Comment by AidanGG** [comment]
Alright, these changes appear to have fixed the docs building.

**Review by ilayn — APPROVED** [review]

**Comment by ilayn** [comment]
Thanks! Now it looks good to me, let's wait for a while in case others want to chime in.

**Comment by ilayn** [comment]
OK let's push this in. Since this is for 1.4 we would still have time to fix it in case I missed a detail. Thanks again @AidanGG, Keep them coming!

And then the obligatory copy/paste response: 

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

**Comment by AidanGG** [comment]
Thanks, @ilayn. I'm thinking about adding some of the routines necessary for tiled Cholesky, which I'll get around to doing once I need them for my own work.

**Pull request merged**